README.md
## pytranscoder
Python wrapper for ffmpeg for batch, concurrent, or clustered transcoding using defined profiles and optional rules engine
for automation.

[Read The Docs](https://pytranscoder.readthedocs.io/en/latest/)


#### News

**Coming soon**

Version 3 will include an alternate means of configuring your setup.  Let's face it, all that YAML is confusing.
The new version will use templating and be far simpler and very much smaller -- all my converted profiles fit on one screen now.
Existing profile support will remain indefinitely for users with more complicated setups, but templates will appeal more to most users.


#### What it is
This script is intended to help automate transcoding for people encoding lots of video.
It is more than a wrapper - it is a workflow and job manager.

There are 2 modes: **local** and **clustered**.  Local mode is the most common usage and is for running this script on the same machine where it is installed.  Cluster mode turns pytranscoder into a remote encoding manager.  In this mode it delegates and manages encode jobs running on multiple hosts.  This requires more advanced configuration and is documented separately in [Cluster.md](https://github.com/mlsmithjr/transcoder/blob/master/Cluster.md)

The remainder of this document focuses on using pytranscoder in local mode.

#### Features:
* Sequential or concurrent transcoding. 
* Concurrent mode allows you to make maximum use of your 
nVidia CUDA-enabled graphics card or Intel accelerated video (QSV)
* Preserves all streams but allows for filtering by audio and subtitle language.
* Configurable transcoding profiles
* Configurable rules and criteria to auto-match a video file to a transcoding profile
* Mixins are profile fragments you can provide at runtime to dynmically change your selected profiles.
* Transcode from a list of files (queue) or all on the command line
* Cluster mode allows use of other machines See [Cluster.md](https://github.com/mlsmithjr/transcoder/blob/master/Cluster.md) for details.
* On-the-fly compression monitoring and optional early job termination if not compressing as expected.
* Experimental support for HandBrakeCLI

#### Requirements

* Linux or MacOS, Windows 10. For Windows, WSL (Ubuntu) recommended.
* latest *ffmpeg* (3.4.3-2 or higher, lower versions may still work)
* nVidia graphics card with latest nVidia CUDA drivers (_optional_)
* Intel CPU with QSV enabled (_optional_)
* Python 3 (3.7 or higher)

#### Support
Please log issues or questions via the github home page for now.

Video Tutorials: [Part 1 - Linux Setup](https://www.youtube.com/watch?v=LHhC_w34Kd0&t=5s), [Part 2 - Usage](https://www.youtube.com/watch?v=Os6UACDAOKA)

#### Installation

There are a few possible ways to install a python app - one of these should work for you.

**Linux**
 The confusion is due to the fact that not all distributions or OS's install pip3 by default. Either way, pytranscoder is available in the **pypi** repo.
```bash
pip3 install --user pytranscoder-ffmpeg
# or...
python3 -m pip install --user pytranscoder-ffmpeg 
```

**Windows (WSL - Ubuntu)**
Windows Subsystem for Linux is the best option, but requires a couple of maintenance steps first if you don't have pip3:
```bash
sudo apt update
sudo apt upgrade
sudo install python3-pip

# now we can install
pip3 install --user pytranscoder-ffmpeg
```
At this point you have a choice - either install ffmpeg for Windows [ffmpeg.exe](https://www.ffmpeg.org) or install in bash as an Ubuntu package. Either will work but there are caveats, or  you could install both and not worry.

* ffmpeg.exe can be run in Windows command shell or from bash but requires special attention when configuring pytranscoder paths.
* ffmpeg apt package can only be run from bash but is a more natural Linux path mapping.

After installing you will find this document and others in $HOME/.local/shared/doc/pytranscoder (on Linux/MacOS)
and in $HOME/AppData/Python/*pythonversion*/shared/doc/pytranscoder** (on Windows). Also available [online](https://github.com/mlsmithjr/transcoder/blob/master/README.md)

#### Upgrading

Whatever method above for installing works for you, just use the --upgrade option to update, ie:
```pip3 install --upgrade pytranscoder-ffmpeg```


#### Operation - Profiles and Rules

>A profile is a named group of *ffmpeg* commandline options to encode a specific way. You can
define all the combinations you use regularly in *transcode.yml* for easy selection later.
At least 1 profile definition is required.

>A rule is a YAML syntax of predefined predicates to allow simple matching on source media details
and relate to a specific profile.  For example, if you transcode 720p differently than 1080p, and still different
than 4k you can set up rules to match those 3 resolutions to a specific transcode profile.
Easy - let the script do the work of selecting the right *ffmpeg* options.

But you aren't required to use rules.  You can specify the profile name on the commandline each
run using the -p option. Or you can define 1 rule that acts as a default (see examples). It's up to you. But using rules is a great way to automate a tedious manual workflow.

When changing or adding profiles and rules it is useful to test them out by running in *--dry-run* mode first, 
which will show you everything that would happen if running for real.

#### Configuration

There is a supplied sample *transcode.yml* config file, or you can download it [here](https://github.com/mlsmithjr/transcoder/blob/master/transcode.yml).  This can be customized all you like, however be
sure to preserve the YAML formatting. Either specify this file on the commandline with the *-y* option
or copy it to your home directory as *.transcode.yml* (default)

There are 3 sections in the file:

#### config - Global configuration information

Sample
```yaml
config:
  default_queue_file:   '/path/to/default/list/of/files/if/none/given'
  ffmpeg:               '/usr/bin/ffmpeg'       # path to ffmpeg for this config
  ssh:                '/usr/bin/ssh'    # used only in cluster mode (optional)
  queues:
    qsv:                1                   # sequential encodes
    cuda:               2                   # maximum of 2 encodes at a time
  colorize:             yes
```

| setting      | purpose |
| ----------- | ----------- |
| default_queue_file    | A queue file is just a text file listing out all the media you want to encode. It is not required, but useful when automating a workflow. You can always indicate a queue file on the command line. This just sets the default, if any.   |
| ffmpeg                | Full path to ffmpeg on this host |
| ssh                   | Full path to ssh on this host |
| queues                | If using concurrency, define your queues here. The queue name is whatever you want. Each name specifies a maximum number of concurrent encoding jobs. If none defined, a default sequential queue is used. |
| colorize     | optional, defaults to "no". If "yes" terminal output will have some color added |

#### profiles - Transcoding profiles (ffmpeg options)

Profiles are used to provide ffmpeg with various options for encoding. One profile definition is required, but mostly likely
you will define multiples for different encoding scenarios.  The name of the profile can be provided on the command line
to select the appropriate one for your needs. Alternatively, you can define rules (see below) to auto-match media with profiles for a less manual encoding workflow.

Sample:
```yaml
profiles:

  # some common, reusable settings to keep things tidy
  common:
    output_options_subtitles:
      - "-c:s copy"
       - "-f matroska"
    output_options_audio:
      - "-c:a copy"
    output_options_video:
      - "-crf 20"
    extension: '.mkv'
    threshold: 20
    threshold_check: 30

  #
  # Sample Intel QSV transcode setup (note to customize options for your environment)
  #
  hevc_qsv:
    include: common
    output_options_video:         # mixin-enabled section - overrides common
      - "-c:v hevc_qsv"
      - "-preset medium"
      - "-qp 21"
      - "-b:v 7M"
    
  #
  # Sample nVidia transcode setup
  #

  hevc_cuda:                  # profile name
      include: common
      input_options:          # ffmpeg input options
        - "-hwaccel cuvid"    # REQUIRED for CUDA
        - "-c:v h264_cuvid"   # hardware decoding too
      output_options_video:   # mixen-enabled - overrides common
        - "-c:v hevc_nvenc"   # REQUIRED for CUDA
        - "-profile:v main"
        - "-preset medium"
      output_options_audio:
        - "-c:a copy"
      
      queue: cuda		# manage this encode in the 'cuda' queue defined globally
      
      # optionally you can filter out audio/subtitle tracks you don't need
      audio:
         exclude_languages:
             - "chi"
             - "spa"
             - "fre"
             - "ger"
         default_language: eng
 
      subtitle:
         exclude_languages:
             - "chi"
             - "spa"
             - "fre"
             - "por"
             - "ger"
             - "jpn"
         default_language: eng
  #
  # This is a mixin, a profile fragment that can be provided on the command line
  # to alter a profile (ie. pytranscoder -p hevc_cuda -m mp3_hq file.mp4)

  mp3_hq:
    output_options_audio:
      - "-c:a mp3lame"
      - "-b:a 384k"

  x264:                            # profile name
      include: common
      input_options: 
      output_options:
        - "-c:v x264"
        
h264_cuda_anime:
    include: common
    input_options:
    output_options:
      - "-c:v h264_nvenc"
      - "-tune animation"
 

CHANGES

### Version History

Version 2.2.7:
    * Fixed streaming host class where spaces in filenames weren't being recognized by Windows

Version 2.2.6:
    * Added new templates which can be used instead of the more complicated profiles

Version 2.2.5:
    * Much overdue doc updates
    * Addition of agent mode as alternative to using ssh.
    * Fixed issue cleaning up temp files on Windows host in streaming mode.

Version 2.2.4:
    * Fixed issue where sometimes the wrong video stream is detected when there are multiple

Version 2.2.3:
    * Minor issue with host name status during clustered transcoding

Version 2.2.1:
    * Bug fix when no mixins given.

Version 2.2:
    * Removed unportable is_mount detection. Now always default to any designated temp drive/path for transcoding.
    * Added mixins as a better alternative to includes (see docs).

Version 2.1.0:
    * Added experimental support for HandBrakeCLI.
    * Requires Python 3.7 or higher
    * Removed Plex integration.
    * Removed Sonarr integration
    * Fixed typo in sample.yml and updated Cuda examples.

Version 2.0.14:
    * Fixed ffprobe output parsing problem - thanks Grant.

Version 2.0.13
    * Added support for ffprobe media analysis as a fallback if normal ffmpeg info dump parsing fails.
    * Fixed a regex issue where certain patterns in a "path" criteria didn't match
    * Tweaks to the sample transcode.yml

Version 2.0.12
    * Fixed bad unlink() reference in threshold cleanup.

Version 2.0.11
    * Fixed bug preventing completed work items from being removed from the queue file.

Version 2.0.10
    * Fixed but on Windows trying to detect mounted filesystem.  Not available under Windows.

Version 2.0.9
    * Added 'fls_path' option to transcode.yml, which allows designating a "fast local storage" path where
      output is written during transcode rather than thrashing your network share/NAS with many small reads
      and writes. Final output is moved to the share when complete.  Only applicable if the input file is on
      a detected network share.

    * Configuration files now require lists to use the multi-line YAML format for consistency (see updated example included)

Version 2.0.8
    * Fixed automap=No recognition problem using non-clustered transcoding.

Version 2.0.7
    * Fixed minor string formatting problem on abort of encode.

Version 2.0.6
    * Suppress summary on --dry-run

Version 2.0.5
    * Fixed another unhandled exception collecting statistics when using non-clustered transcoding.

Version 2.0.4
    * Fixed unhandled exception collecting statistics when using non-clustered transcoding.

Version 2.0.3
    * Prevent error message when remote host health testing times out.
    * Fixed erroneous "queue file not found" message.

Version 2.0.2
    * Added summary output of full job with elapsed time
    * Fixed output bug which failed to report proper filename on termination of encode job.

Version 2.0.1
    * Added "automap" setting to Global and Profile sections to expressly enable or disable mapping (default is ENABLED).
    * Added include_languages to "audio" and "subtitle" profile sections.  Use Include or Exclude, not both.  If you
      use both then Includes will take precedence.
    * Enabled audio and subtitle filtering in cluster mode too.
    * Fixed bug in inverted vcodec check (ex. vcodec !h264)
    * Fixed bug where cluster mode wasn't removing completed jobs from the queue file.
    * Been running reliably and well for months now - time for a version bump to 2.0!
    * Added a ctrl-c (sigint) handler to prevent threads from removing original media during termination.
      This seems to be a python 3.7 thing.

Version 1.5.12
    * Merging of input_options and output_options supported at the option level. For example, if the included
      profiles has "-c:a copy" and the including profile has "-c:a libmp3lame", the including profile option will
      take precedence.

Version 1.5.11
    * Default behavior is to now auto-map all audio and subtitle streams during transcode.
        May be disabled at the global or profile level with  automap: no
    * More unit test fixes and improvements.
    * Updated sample transcode.yml for better feature reference

Version 1.5.10
    * Fixed broken unit tests from previous release
    * Fixed rule predicates matching on runtime after conversion to seconds-level detail.

Version 1.5.9
    * Added new optional "include" directive in profiles to help reduce profile repetition.
    * Fixed error message when input file not found.
    * Fixed divide-by-zero error when video runtime < 1 minute. Calculations how include seconds rather than just whole minutes

Version 1.5.8
    * Updated yaml loader to support latest pyyaml library.

Version 1.5.7

transcodertests.py

import unittest
import os
from typing import Dict
from unittest import mock

from pytranscoder.cluster import RemoteHostProperties, Cluster, StreamingManagedHost
from pytranscoder.config import ConfigFile
from pytranscoder.ffmpeg import status_re, FFmpeg
from pytranscoder.media import MediaInfo
from pytranscoder.profile import Profile
from pytranscoder.transcode import LocalHost
from pytranscoder.utils import files_from_file, get_local_os_type, calculate_progress, dump_stats, is_exceeded_threshold


class TranscoderTests(unittest.TestCase):

    @staticmethod
    def make_media(path, vcodec, res_width, res_height, runtime, source_size, fps, colorspace,
                   audio, subtitle) -> MediaInfo:
        info = {
            'path': path,
            'vcodec': vcodec,
            'stream': 0,
            'res_width': res_width,
            'res_height': res_height,
            'runtime': runtime,
            'filesize_mb': source_size,
            'fps': fps,
            'colorspace': colorspace,
            'audio': audio,
            'subtitle': subtitle
        }
        return MediaInfo(info)

    def test_stream_map_all(self):
        with open('tests/ffmpeg3.out', 'r') as ff:
            info = MediaInfo.parse_ffmpeg_details('/dev/null', ff.read())
            setup = ConfigFile(self.get_setup())
            p = setup.get_directive('qsv')
            streams = info.stream_map(p)
            self.assertEqual(len(streams), 2, 'expected -map 0')

    def test_stream_exclude(self):
        with open('tests/ffmpeg3.out', 'r') as ff:
            info = MediaInfo.parse_ffmpeg_details('/dev/null', ff.read())
            setup = ConfigFile(self.get_setup())
            p = setup.get_directive('excl_test_1')
            streams = info.stream_map(p)
            self.assertEqual(len(streams), 12, 'expected 6 streams (12 elements)')

    def test_stream_reassign_default(self):
        with open('tests/ffmpeg4.out', 'r') as ff:
            info = MediaInfo.parse_ffmpeg_details('/dev/null', ff.read())
            setup = ConfigFile(self.get_setup())
            p = setup.get_directive('excl_test_2')
            streams = info.stream_map(p)
            self.assertEqual(len(streams), 8, 'expected 4 streams (8 elements)')

#    def test_hook(self):
#        transcode.manage_hook("rule_hook_ex.py")
#        info = TranscoderTests.make_media("/some/path/to/anime/content", "x264", 1080, 720, 45, 3000, 25, None, [], [])
#        profile = try_hook(info, is_testing=True)
#        transcode.manage_hook(None)
#        self.assertEqual(profile.name, "anime", "Expected to match anime")
#        expected = sorted(["-f matroska", "-c:a copy", "-c:s copy", "-c:v hevc_nvenc", "-profile:v main",
#                           "-preset medium", "-crf 20"])
#        actual = sorted(profile.output_options.as_list())
#        self.assertEqual(actual, expected, "Output options mismatch")

    def test_profile_merge(self):
        p1 = Profile("p1", {"one": 1, "two": 2, "input_options": ["three", "four", "five"]})
        p2 = Profile("p2", {"one": 11, "six": 6, "input_options": ["seven", "four"]})
        p2.include(p1)
        self.assertEqual(p2.get("one"), 11, 'expected 11 for "one"')
        self.assertEqual(p2.get("two"), 2, 'expected 2 for "two"')
        self.assertEqual(p2.get("six"), 6, 'expected 6 for "six"')
        op2 = sorted(p2.input_options_list.as_list())
        expected = sorted(["three", "four", "five", "seven"])
        self.assertEqual(op2, expected, "Unexpected input_options merger")

    def test_progress(self):
        info = TranscoderTests.make_media(None, None, None, 1080, 90 * 60, 2300, 25, None, [], [])
        stats = {'size': 1225360000, 'time': 50 * 60}
        done, comp = calculate_progress(info, stats)
        self.assertEqual(done, 55, 'Expected 55% done')
        self.assertEqual(comp, 6, 'Expected 6% compression')

    def test_ffmpeg_status_regex(self):
        sample = 'frame=  307 fps= 86 q=-0.0 size=    3481kB time=00:00:13.03 bitrate=2187.9kbits/s speed=3.67x   \n'
        match = status_re.match(sample)
        self.assertIsNotNone(match, 'no ffmpeg status match')
        self.assertTrue(len(match.groups()) == 5, 'Expected 5 matches')

    def test_loadconfig(self):
        config = ConfigFile('config-samples/transcode.yml')
        self.assertIsNotNone(config.settings, 'Config object not loaded')
        self.assertEqual(len(config.queues), 2, 'Expected 2 queues')

    def test_loadqueue(self):
        testpath = '/tmp/transcode_parsertest_loadqueue.tmp'
        with open(testpath, 'w') as t:
            t.write('one\ntwo\nthree')
        files = files_from_file(testpath)
        self.assertTrue(len(files) == 3, 'did not load file queue properly')
        os.remove(testpath)

    def test_mediainfo(self):
        with open('tests/ffmpeg.out', 'r') as ff:
            info = MediaInfo.parse_ffmpeg_details('/dev/null', ff.read())
            self.assertIsNotNone(info)
            self.assertEqual(info.vcodec, 'h264')
            self.assertEqual(info.res_width, 1280)
            self.assertEqual(info.fps, 23)
            self.assertEqual(info.runtime, (2 * 3600) + (9 * 60) + 38)
            self.assertEqual(info.path, '/dev/null')
            self.assertEqual(info.colorspace, 'yuv420p')

    def test_mediainfo2(self):
        with open('tests/ffmpeg2.out', 'r') as ff:
            info = MediaInfo.parse_ffmpeg_details('/dev/null', ff.read())
            self.assertIsNotNone(info)
            self.assertEqual(info.vcodec, 'h264')
            self.assertEqual(info.res_width, 1920)
            self.assertEqual(info.fps, 24)
            self.assertEqual(info.runtime, (52 * 60) + 49)
            self.assertEqual(info.path, '/dev/null')
            self.assertEqual(info.colorspace, 'yuv420p')

    def test_mediainfo3(self):
        with open('tests/ffmpeg3.out', 'r') as ff:
            info = MediaInfo.parse_ffmpeg_details('/dev/null', ff.read())
            self.assertIsNotNone(info)
            self.assertEqual(info.vcodec, 'hevc')
            self.assertEqual(info.res_width, 3840)
            self.assertEqual(info.fps, 23)
            self.assertEqual(info.runtime, (2 * 3600) + (5 * 60) + 53)
            self.assertEqual(info.path, '/dev/null')
            self.assertEqual(info.colorspace, 'yuv420p10le')

    def test_include_overides(self):
        setup = ConfigFile(self.get_setup())
        p: Profile = setup.get_directive('hevc_cuda_8bit')
        self.assertEqual(p.threshold, 0, 'Threshold should be 0')
        self.assertEqual(p.threshold_check, 100, 'Threshold check should be 100')
        self.assertIn("-cq:v 21", p.output_options_list.as_list(), 'Expected -cq:v 21')
        self.assertIn('-pix_fmt yuv420p', p.output_options_list.as_list(), 'expected pix_fmt')

    def test_default_profile(self):
        info = TranscoderTests.make_media(None, None, None, 720, 45, 3000, 25, None, [], [])
        config = ConfigFile(self.get_setup())
        rule = config.match_rule(info)
        self.assertIsNotNone(rule, 'expected to match a rule')
        self.assertEqual(rule.profile, 'hevc_cuda')
        self.assertEqual(rule.name, 'default')

    def test_skip_profile(self):
        with open('tests/ffmpeg.out', 'r') as ff:
            info = MediaInfo.parse_ffmpeg_details('/dev/null', ff.read())
            info.filesize_mb = 499
            config = ConfigFile(self.get_setup())
            rule = config.match_rule(info)
            self.assertIsNotNone(rule, 'Expected rule match')
            self.assertTrue(rule.is_skip(), 'Expected a SKIP rule')

    def test_rule_match(self):
        info = TranscoderTests.make_media(None, None, None, 1080, 45, 2300, None, None, [], [])
        config = ConfigFile(self.get_setup())
        rule = config.match_rule(info)
        self.assertIsNotNone(rule, 'Expected a matched profile')

    def test_loc_os(self):
        self.assertNotEqual(get_local_os_type(), 'unknown', 'Expected other than "unknown" as os type')

    def test_path_substitutions(self):
        config: Dict = self.get_setup()
        props = RemoteHostProperties('m1', config['config']['clusters']['cluster1']['m1'])
        intest, outtest = props.substitute_paths('/volume2/test.in', '/volume2/test.out')
        self.assertEqual(intest, '/media/test.in', 'Path substitution failed on input path')
        self.assertEqual(outtest, '/media/test.out', 'Path substitution failed on output path')

    def test_local_host_setup(self):
        config: Dict = self.get_setup()
        host = LocalHost(ConfigFile(config))
        self.assertEqual(len(host.queues), 3, 'Expected 3 queues configured')

    def test_threshold_calculation(self):
        src = 1_234_567_890
        dest = 1_000_000_000
        threshold = 20
        result = is_exceeded_threshold(threshold, src, dest)
        self.assertFalse(result, "Expected threshold to be false")

    @staticmethod
    def get_setup():
        setup = {
            'config': {
                'ffmpeg': '/usr/bin/ffmpeg',
                'hbcli': '/usr/bin/HandBrakeCLI',
                'queues': {
                    'q1': 1,
                    'q2': 2
                },
                'clusters': {
                    'cluster1': {
                        'm1': {
                            'type': 'mounted',
                            'ip': '127.0.0.1',
                            'user': 'mark',
                            'os': 'linux',
                            'ffmpeg': '/usr/bin/ffmpeg',
                            'path-substitutions': [
                                '/v2/ /m2/',
                                '/volume2/ /media/'
                            ],
                            'profiles': ['hevc_cuda'],
                            'queues': {'q2': 2},
                            'status': 'enabled',
                        },
                        'workstation': {
                            'os': 'linux',
                            'type': 'local',
                            'ip': '192.168.2.63',
                            'ffmpeg': '/usr/bin/ffmpeg',
                            'status': 'enabled',
                        },
                        'm2': {
                            'type': 'streaming',
                            'ip': '127.0.0.1',
                            'os': 'linux',
                            'user': 'mark',
                            'ffmpeg': '/usr/bin/ffmpeg',
                            'working_dir': '/tmp/pytranscode-remote',
                            'profiles': ['qsv'],
                            'queues': {'q3': 1},
                            'status': 'enabled',
                        },
                    },
                },
            },
            "profiles": {
                "hq": {
                    "output_options": ["-c:v", "copy", "-c:a", "copy", "-c:s", "copy", "-f", "matroska"],
                    "threshold": 1,
                    "extension": ".mkv",
                },
                "hevc_cuda": {
                    "include": "hq",
                    "input_options": None,
                    "output_options": ["-threads 4"],
                    "extension": ".mkv",
                    "queue": "q2",
                    "audio": {
                        "include_languages": ["eng"],
                        "default_language": "eng",
                    },
                    "subtitle": {
                        "include_languages": ["eng"],
                        "default_language": "eng",
                    }
                },
                "hevc_cuda_8bit": {
                    "include": "hq",
                    "output_options": [
                        "-cq:v 21",
                        "-b:v 2000K",
                        "-maxrate:v 4M",
                        "-pix_fmt yuv420p"
                    ],
                    "threshold": 0,
                    "threshold_check": 100
                },
                "excl_test_1": {
                    "include": "hq",
                    "input_options": None,
                    "output_options": ["-threads 4"],
                    "extension": ".mkv",
                    "queue": "q2",
                    "audio": {
                        "exclude_languages": ["spa"],
                        "default_language": "eng",
                    },
                    "subtitle": {
                        "exclude_languages": ["spa"],
                        "default_language": "eng",
                    }
                },
                "excl_test_2": {
                    "include": "hq",
                    "input_options": None,
                    "output_options": ["-threads 4"],
                    "extension": ".mkv",
                    "queue": "q2",
                    "audio": {
                        "exclude_languages": ["eng"],
                        "default_language": "chi",
                    },
                    "subtitle": {
                        "exclude_languages": ["eng"],
                        "default_language": "chi",
                    }
                },
                "qsv": {
                    "input_options": None,
                    "output_options": ["-c:v copy -c:a copy"],
                    "extension": ".mkv",
                    "queue": "q3",
                },
                "vintage_tv": {
                    "input_options": None,
                    "output_options": ["-c:v copy -c:a copy"],
                    "extension": ".mp4",
                },
                "hbcli_qsv": {
                    "input_options": None,
                    "output_options": ["-f av_mkv -q 20.0 -B 256 -e qsv_h265"],
                    "extension": ".mkv",
                    "queue": "qsv"
                }
            },
            "rules": {
                'vintage tv': {
                    'profile': 'vintage_tv',
                    'criteria': {
                        'filesize_mb': '<500',
                        'res_height': '<500',
                        'runtime': '<60',
                    },
                },
                'too small': {
                    'profile': 'SKIP',
                    'criteria': {
                        'filesize_mb': '<500'
                    }
                },
                'small enough already': {
                    'profile': 'SKIP',
                    'criteria': {
                        'filesize_mb': '<2500',
                        'res_height': '720-1081',
                        'runtime': '30-65'
                    }
                },
                'feature-length treat better': {
                    'profile': 'qsv',
                    'criteria': {
                        'runtime': '>90'
                    },
                },
                'default': {
                    'profile': 'hevc_cuda',
                    'criteria': {
                        'vcodec': '!hevc'
                    }
                }
            }
        }
        return setup

    @mock.patch.object(FFmpeg, 'run_remote')
    @mock.patch('pytranscoder.cluster.filter_threshold')
    @mock.patch('pytranscoder.cluster.os.rename')
    @mock.patch('pytranscoder.cluster.os.remove')
    @mock.patch.object(FFmpeg, 'fetch_details')
    def test_cluster_match_default_rule(self, mock_ffmpeg_details, mock_os_rename, mock_os_remove,
                                        mock_filter_threshold, mock_run_remote):

        setup = ConfigFile(self.get_setup())

        #
        # setup all mocks
        #
        mock_run_remote.return_value = 0
        mock_filter_threshold.return_value = True
        mock_os_rename.return_value = None
        mock_os_remove.return_value = None
        info = TranscoderTests.make_media('/dev/null', 'x264', 1920, 1080, 45 * 60, 3200, 24, None, [], [])
        mock_ffmpeg_details.return_value = info

        #
        # configure the cluster, add the job, and run
        #
        cluster = self.setup_cluster1(setup)
        qname, job = cluster.enqueue('/dev/null.mp4', None)
        self.assertEqual(qname, 'q2', 'Job placed in wrong queue')
        self.assertEqual(job.profile_name, 'hevc_cuda', 'Rule matched to wrong profile')

        cluster.testrun()
        for host in cluster.hosts:
            if host.hostname == 'm1' and len(host._complete) > 0:
                filename, elapsed = host.completed.pop()
                self.assertEqual('/dev/null.mp4', filename, 'Completed filename missing from assigned host')
                break

    @staticmethod
    def setup_cluster1(config) -> Cluster:
        cluster_config = config.settings['clusters']
        cluster = Cluster('cluster1', cluster_config['cluster1'], config, config.ssh_path)
        return cluster

    @mock.patch.object(FFmpeg, 'run_remote')
    @mock.patch('pytranscoder.cluster.filter_threshold')
    @mock.patch('pytranscoder.cluster.os.rename')
    @mock.patch('pytranscoder.cluster.os.remove')
    @mock.patch.object(FFmpeg, 'fetch_details')
    def test_cluster_match_skip(self, mock_ffmpeg_details, mock_os_rename, mock_os_remove,
                                mock_filter_threshold, mock_run_remote):

        setup = ConfigFile(self.get_setup())

        #
        # setup all mocks
        #
        mock_run_remote.return_value = 0
        mock_filter_threshold.return_value = True
        mock_os_rename.return_value = None
        mock_os_remove.return_value = None
        info = TranscoderTests.make_media('/dev/null', 'x264', 1920, 1080, 60 * 60, 1800, 24, None, [], [])
        mock_ffmpeg_details.return_value = info

        #
        # configure the cluster, add the job, and run
        #
        cluster = self.setup_cluster1(setup)
        qname, job = cluster.enqueue('/dev/null.mp4', None)
        self.assertEqual(qname, None, 'Expected to skip')

    @mock.patch.object(FFmpeg, 'run_remote')
    @mock.patch('pytranscoder.cluster.filter_threshold')
    @mock.patch('pytranscoder.cluster.os.rename')
    @mock.patch('pytranscoder.cluster.os.remove')
    @mock.patch.object(MediaInfo, 'parse_ffmpeg_details')
    @mock.patch('pytranscoder.cluster.run')
    @mock.patch('pytranscoder.cluster.shutil.move')
    @mock.patch.object(FFmpeg, 'fetch_details')
    @mock.patch.object(StreamingManagedHost, 'run_process')
    def test_cluster_streaming_host(self, mock_run_proc, mock_ffmpeg_fetch, mock_move, mock_run, mock_info_parser,
                                    mock_os_rename, mock_os_remove,
                                    mock_filter_threshold, mock_run_remote):

        setup = ConfigFile(self.get_setup())

        #
        # setup all mocks
        #
        mock_run.return_value = 0, 'ok'
        mock_run_remote.return_value = 0
        mock_move.return_value = 0
        mock_filter_threshold.return_value = True
        mock_os_rename.return_value = None
        mock_os_remove.return_value = None
        info = TranscoderTests.make_media('/dev/null', 'x264', 1920, 1080, 110 * 60, 3000, 24, None, [], [])
        mock_info_parser.return_value = info
        mock_ffmpeg_fetch.return_value = info
        #
        # configure the cluster, add the job, and run
        #
        cluster = self.setup_cluster1(setup)
        qname, job = cluster.enqueue('/dev/null.mp4', None)
        self.assertEqual(qname, 'q3', 'Job placed in wrong queue')
        self.assertEqual(job.profile_name, 'qsv', 'Rule matched to wrong profile')

        cluster.testrun()
        for host in cluster.hosts:
            if host.hostname == 'm2' and len(host._complete) > 0:
                filename, elapsed = host.completed.pop()
                self.assertEqual('/dev/null.mp4', filename,
                                  'Completed filename missing from assigned host')
                break

    @mock.patch.object(FFmpeg, 'run')
    @mock.patch('pytranscoder.cluster.filter_threshold')
    @mock.patch('pytranscoder.cluster.os.rename')
    @mock.patch('pytranscoder.cluster.os.remove')
    @mock.patch.object(MediaInfo, 'parse_ffmpeg_details')
    @mock.patch.object(FFmpeg, 'fetch_details')
    def test_cluster_match_default_queue(self, mock_ffmpeg_details, mock_info_parser, mock_os_rename, mock_os_remove,
                                         mock_filter_threshold, mock_run_remote):

        setup = ConfigFile(self.get_setup())

        #
        # setup all mocks
        #
        mock_run_remote.return_value = 0
        mock_filter_threshold.return_value = True
        mock_os_rename.return_value = None
        mock_os_remove.return_value = None
        info = TranscoderTests.make_media('/dev/null', 'x264', 500, 480, 30 * 60, 420, 24, None, [], [])
        mock_info_parser.return_value = info
        mock_ffmpeg_details.return_value = info

        #
        # configure the cluster, add the job, and run
        #
        cluster = self.setup_cluster1(setup)
        qname, job = cluster.enqueue('/dev/null.mp4', None)
        self.assertEqual(qname, '_default', 'Job placed in wrong queue')
        self.assertEqual(job.profile_name, 'vintage_tv', 'Rule matched to wrong profile')

        cluster.testrun()
        for host in cluster.hosts:
            if host.hostname == 'workstation' and len(host._complete) > 0:
                dump_stats(host._complete)
                filename, elapsed = host.completed.pop()
                self.assertEqual('/dev/null.mp4', filename, 'Completed filename missing from assigned host')
                break


if __name__ == '__main__':
    unittest.main()
